flows/inputs/base.go

```diff
@@ -13,7 +13,7 @@ import (
 	"github.com/pkg/errors"
 )
 
-type readFunc func(session flows.Session, data json.RawMessage) (flows.Input, error)
+type readFunc func(flows.SessionAssets, json.RawMessage, func(assets.Reference)) (flows.Input, error)
```

Comment:
The `func(assets.Reference)` thing is a bit opaque. Would be better to make this a type so that it is a bit more descriptive. (the contact modifier application method suffers the same)